### Задание 1
#### Импортируйте библиотеки pandas и numpy.


In [1]:
import pandas as pd
import numpy as np

#### Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. Создайте датафреймы X и y из этих данных.

In [6]:
from sklearn.datasets import load_boston
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR'])

In [8]:
data = boston["data"]
data.shape

(506, 13)

In [9]:
feature_names = boston["feature_names"]
feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [11]:
target = boston["target"]
target[:10]

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9])

In [10]:
X = pd.DataFrame(data, columns=feature_names)
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [12]:
y = pd.DataFrame(target, columns=["price"])
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
price    506 non-null float64
dtypes: float64(1)
memory usage: 4.0 KB


#### Разбейте эти датафреймы на тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

#### Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [15]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()

#### Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [16]:
lr.fit(X_train, y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [17]:
y_pred = lr.predict(X_test)
y_pred.shape

(152, 1)

In [18]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})
check_test.head(10)

y_test     y_pred
173    23.6  28.655072
274    32.4  36.502090
491    13.6  15.418775
72     22.8  25.412343
452    16.1  18.845041
76     20.0  23.151931
316    17.8  17.390458
140    14.0  14.070685
471    19.6  23.041064
500    16.8  20.599791

In [20]:
check_test["error"] = check_test["y_pred"] - check_test["y_test"]
check_test.head()

y_test     y_pred     error
173    23.6  28.655072  5.055072
274    32.4  36.502090  4.102090
491    13.6  15.418775  1.818775
72     22.8  25.412343  2.612343
452    16.1  18.845041  2.745041

#### Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [21]:
from sklearn.metrics import r2_score
r2_score(check_test["y_test"], check_test["y_pred"])

0.7109203586326303

### Задание 2

#### Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble. Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

In [25]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators = 1000, max_depth = 42)
model

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=42,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

#### Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression, но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0], чтобы получить из датафрейма одномерный массив Numpy,так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.



In [26]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=42,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=1000, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

#### Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

In [27]:
y_pred = model.predict(X_test)
y_pred.shape

(152,)

In [28]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head(10)

y_test   y_pred
173    23.6  22.7655
274    32.4  30.9810
491    13.6  16.3220
72     22.8  23.8659
452    16.1  16.9372
76     20.0  21.7536
316    17.8  19.8659
140    14.0  14.6718
471    19.6  21.2884
500    16.8  20.8203

In [29]:
r2_score(check_test["y_test"], check_test["y_pred"])

0.8751641212223473

При использовании LinearRegression коэффициент детерминации R2 = 0.7109203586326303, а в случае использования RandomForestRegressor R2 = 0.8751641212223473. Коэффициент детерминации должен быть хотя бы не меньше 50%,  модели с коэффициентом детерминации выше 80 % можно признать достаточно хорошими.